uat/testing-features/src/main/java/com/aws/greengrass/aws/S3Bucket.java

```diff
@@ -0,0 +1,57 @@
+package com.aws.greengrass.aws;
+
+import com.amazonaws.services.s3.model.Bucket;
+import com.amazonaws.services.s3.model.ObjectListing;
+import com.aws.greengrass.testing.resources.AWSResource;
+import lombok.Data;
+import lombok.extern.log4j.Log4j2;
+
+import com.amazonaws.services.s3.AmazonS3;
+import com.amazonaws.services.s3.model.S3ObjectSummary;
+
+import static com.aws.greengrass.aws.IotShadowResources.randomName;
+
+@Data
+@Log4j2
+public class S3Bucket implements AWSResource<AmazonS3> {
+    boolean deleted = false;
+    private String bucketName;
+
+    public S3Bucket(String bucketName) {
+        this.bucketName = bucketName;
+    }
+
+    public static S3Bucket create(AmazonS3 amazonS3, S3BucketSpec spec) {
+        if (spec.bucketName == null) {
+            spec.setBucketName(randomName());
+        }
+        log.atDebug().log("Creating a S3 Bucket {}", spec.bucketName);
+        Bucket b = amazonS3.createBucket(spec.bucketName);
+        return new S3Bucket(spec.bucketName);
+    }
+
+    @Override
+    public void remove(AmazonS3 amazons3) {
+        if (deleted) {
+            return;
+        }
+        log.atDebug().log("Removing S3 Bucket {}", bucketName);
+        // Assuming that S3 bucket is not versioned in this case
+        // removing objects from bucket
+        ObjectListing object_listing = amazons3.listObjects(bucketName);
+        while (true) {
+            for (S3ObjectSummary summary : object_listing.getObjectSummaries()) {
+                amazons3.deleteObject(bucketName, summary.getKey());
+            }
+
+            // more object_listing to retrieve?
+            if (object_listing.isTruncated()) {
+                object_listing = amazons3.listNextBatchOfObjects(object_listing);
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

This code is manually paginating results. You could use the SDK to automatically paginate these results instead. SDK operation name: [S3Objects.withPrefix](https://docs.aws.amazon.com/AWSJavaSDK/latest/javadoc/com/amazonaws/services/s3/iterable/S3Objects.html#withPrefix-com.amazonaws.services.s3.AmazonS3-java.lang.String-java.lang.String-).
 The list of [S3ObjectSummary](https://docs.aws.amazon.com/AWSJavaSDK/latest/javadoc/com/amazonaws/services/s3/model/S3ObjectSummary.html)s will be fetched lazily, a page at a time, as they are needed. The size of the page can be controlled with the [withBatchSize(int)](https://docs.aws.amazon.com/AWSJavaSDK/latest/javadoc/com/amazonaws/services/s3/iterable/S3Objects.html#withBatchSize-int-) method.